### PR TITLE
add scale factor to Geogrid and autoRIFT-estimated velocity products

### PR DESCRIPTION
**Pull request by leiyangleon**
1. added a new 2-band raster output "window_scale_factor.tif" in Geogrid and fixed a few misuses of pixel size in image vs. map coordinates.
2. incorporated the new Geogrid output "window_scale_factor.tif" in autoRIFT for producing the velocity products
3. transferred scale factor (2-band raster) to netcdf packaging in case you want to package them as well in the final netcdf file 

**NOTE**: Please note I have tested the new Geogrid output "window_scale_factor.tif" as well as the autoRIFT velocity product "velocity.tif" for one S1 and one L8 image pair. The results look reasonable, with the value of scale factor being very close to 1 (e.g. 1.0001 or 0.9999). The velocity only has changes at the 1st or 2nd decimal compared to the original case. You might want to do more tests before merging the PR to double check.

### CODE REVIEW

**Comment by leiyangleon** [comment]
> * added a new 2-band raster output "window_scale_factor.tif" in Geogrid and fixed a few misuses of pixel size in image vs. map coordinates.
> * incorporated the new Geogrid output "window_scale_factor.tif" in autoRIFT for producing the velocity products
> * transferred scale factor (2-band raster) to netcdf packaging in case you want to package them as well in the final netcdf file

File: geo_autoRIFT/geogrid/src/geogridOptical.cpp [hunk 1]
```diff
@@ -1040,8 +1077,8 @@ void geoGridOptical::geogridOptical()
                         }
                         else
                         {
-                            raster11[jj] = std::round(dot_C(vel,xunit)*dt/XSize/365.0/24.0/3600.0*1);
-                            raster22[jj] = std::round(dot_C(vel,yunit)*dt/YSize/365.0/24.0/3600.0*1);
+                            raster11[jj] = std::round(dot_C(vel,xunit)*dt/norm_C(xdiff)/365.0/24.0/3600.0*1);
+                            raster22[jj] = std::round(dot_C(vel,yunit)*dt/norm_C(ydiff)/365.0/24.0/3600.0*1);
```
Comment (alex-s-gardner) [inline geo_autoRIFT/geogrid/src/geogridOptical.cpp:1081]: Are raster11 and raster22 used for optical processing? Why did you need to change from grid size to norm?

File: geo_autoRIFT/geogrid/src/geogridOptical.cpp [hunk 2]
```diff
@@ -1065,6 +1102,10 @@ void geoGridOptical::geogridOptical()
                         raster2b[jj] = nodata_out;
                     }
                     
+                    // range/azimuth distance to DEM distance scale factor
+                    sf_raster1[jj] = norm_C(xdiff) / std::abs(XSize);
+                    sf_raster2[jj] = norm_C(ydiff) / std::abs(YSize);
```
Comment (alex-s-gardner) [inline geo_autoRIFT/geogrid/src/geogridOptical.cpp:1107]: OK, so in this case we now have scale factors but raster1a (1b,2a,2b) remain unmodified. So i guess the scale is still removed from the transformation matrix but now we need to multiply the rotation matrix by the scale factors (sf_raster1 (2)) when transforming the data. OK, I see the addition in netcdf_output.py

File: geo_autoRIFT/geogrid/src/geogrid.cpp [hunk 3]
```diff
@@ -1255,12 +1291,17 @@ void geoGrid::geogrid()
...
+                        raster1b[jj] = -normal[2]/(dt/norm_C(da)/365.0/24.0/3600.0)*(normal[2]*los[1]-normal[1]*los[2])/((normal[2]*los[0]-normal[0]*los[2])*(normal[2]*temp[1]-normal[1]*temp[2])-(normal[2]*temp[0]-normal[0]*temp[2])*(normal[2]*los[1]-normal[1]*los[2]));
                         raster2a[jj] = -normal[2]/(dt/dr/365.0/24.0/3600.0)*(normal[2]*temp[0]-normal[0]*temp[2])/((normal[2]*los[0]-normal[0]*los[2])*(normal[2]*temp[1]-normal[1]*temp[2])-(normal[2]*temp[0]-normal[0]*temp[2])*(normal[2]*los[1]-normal[1]*los[2]));
-                        raster2b[jj] = normal[2]/(dt/norm_C(alt)/365.0/24.0/3600.0)*(normal[2]*los[0]-normal[0]*los[2])/((normal[2]*los[0]-normal[0]*los[2])*(normal[2]*temp[1]-normal[1]*temp[2])-(normal[2]*temp[0]-normal[0]*temp[2])*(normal[2]*los[1]-normal[1]*los[2]));
+                        raster2b[jj] = normal[2]/(dt/norm_C(da)/365.0/24.0/3600.0)*(normal[2]*los[0]-normal[0]*los[2])/((normal[2]*los[0]-normal[0]*los[2])*(normal[2]*temp[1]-normal[1]*temp[2])-(normal[2]*temp[0]-normal[0]*temp[2])*(normal[2]*los[1]-normal[1]*los[2]));
                     }
```
Comment (alex-s-gardner) [inline geo_autoRIFT/geogrid/src/geogrid.cpp:1305]: Why were raster1b/2b modified?

File: geo_autoRIFT/geogrid/src/geogrid.cpp [hunk 4]
```diff
@@ -1270,12 +1311,13 @@ void geoGrid::geogrid()
...
                     
-                    for(int pp=0; pp<3; pp++)
-                    {
-                        targXYZ[pp] -= xyz[pp];
-                    }
+                    // range/azimuth pixel displacement to range/azimuth velocity
                     raster1c[jj] = dr/dt*365.0*24.0*3600.0*1;
-                    raster2c[jj] = norm_C(targXYZ)/dt*365.0*24.0*3600.0*1;
+                    raster2c[jj] = norm_C(da)/dt*365.0*24.0*3600.0*1;
```
Comment (alex-s-gardner) [inline geo_autoRIFT/geogrid/src/geogrid.cpp:1316]: And why doe raster1c/2c get changed here as well... I ask because if there are changes to raster1a,b,c/2a,bc then we can't simply apply the scale factors (sf_raster1/2) to correct previously run data.

**Review by alex-s-gardner — COMMENTED** [review]
I have some questions on modification to raster1a,b,c,2a,b,c

File: geo_autoRIFT/geogrid/src/geogrid.cpp [hunk 5]
```diff
@@ -1245,7 +1281,7 @@ void geoGrid::geogrid()
                         }
                         else
                         {
-                            raster11[jj] = std::round(dot_C(vel,los)*dt/dr/365.0/24.0/3600.0*1);
+                            raster11[jj] = std::round(dot_C(vel,los)*dt/norm_C(drpos)/365.0/24.0/3600.0*1);
```
[the remark this answers is not in the document]
Comment (leiyangleon) [reply geo_autoRIFT/geogrid/src/geogrid.cpp:1284]: @alex-s-gardner that is right. They are not related to scale distortion. When we ignored the scale distortion problem, there was a misuse of pixel size in image CS vs. pixel size in map projection CS. While introducing the scale factor, I realized this inconsistency and the pixel sizes are now corrected as well.

File: geo_autoRIFT/geogrid/src/geogridOptical.cpp (hunk 1 above)
Comment (leiyangleon) [reply geo_autoRIFT/geogrid/src/geogridOptical.cpp:1081]: @alex-s-gardner raster11 and raster22 are downstream search pixel distances, so not related to the conversion matrix. Grid size (XSize and YSize) was defined in image CS (input projection for optical), while norm(xdiff) and norm(ydiff) are transformed pixel sizes in map projection CS (output projection for optical). Again this is related to calculating the downstream search pixel distances: when vel and xunit/yunit are all defined in output projection, the pixel sizes should be also in that CS rather than the input one.

File: geo_autoRIFT/geogrid/src/geogridOptical.cpp (hunk 2 above)
Comment (leiyangleon) [reply geo_autoRIFT/geogrid/src/geogridOptical.cpp:1107]: @alex-s-gardner you are right about the workflow.

File: geo_autoRIFT/geogrid/src/geogrid.cpp (hunk 4 above)
Comment (leiyangleon) [reply geo_autoRIFT/geogrid/src/geogrid.cpp:1316]: @alex-s-gardner 
- raster1c and raster2c are range/azimuth pixel displacement to range/azimuth velocity conversion coefficients. Same reason as above: correction of pixel size misuse. 
- I understand your concern. But by any means, we have to rerun almost everything. For calculating the scale factors, we need transformed displacement vector in map projection CS (norm(xdiff/ydiff) for optical and norm(drpos/alt) for radar), by re-running Geogrid (which is also required to correct the pixel size misuse in conversion matrix calculations). Geogrid takes the ISCE files as input. Since all the ISCE files were removed after processing, we have to re-run ISCE. For autoRIFT, if @jhkennedy Joe saved that intermediate netcdf file "autoRIFT_intermediate.nc", we don't need to rerun autoRIFT but use that file for re-packaging; otherwise, autoRIFT needs to be re-ran as well.

**Review by jhkennedy — APPROVED** [review]
